phpunit/annotations/class-annotation-caps-test.php

```diff
@@ -0,0 +1,321 @@
+<?php
+/**
+ * WP_Annotation_Caps Tests
+ *
+ * @package gutenberg
+ */
+
+require_once dirname( __FILE__ ) . '/includes/class-annotation-test-objects.php';
+require_once dirname( __FILE__ ) . '/includes/class-annotation-test-utils.php';
+
+/**
+ * Tests annotation capabilities.
+ */
+class Annotation_Caps_Test extends WP_UnitTestCase {
+	/**
+	 * Magic first-access properties.
+	 *
+	 * @param  string $property Property name.
+	 * @return mixed            Property value.
+	 */
+	public function __get( $property ) {
+		if ( 'objects' === $property ) {
+			$this->objects = new Annotation_Test_Objects( $this );
+			return $this->objects;
+		} elseif ( 'utils' === $property ) {
+			$this->utils = new Annotation_Test_Utils( $this );
+			return $this->utils;
+		}
+		return parent::__get( $property );
+	}
+
+	/**
+	 * Check that we have necessary hooks.
+	 */
+	public function test_has_hooks() {
+		$this->assertNotEmpty( has_filter( 'map_meta_cap', 'WP_Annotation_Caps::on_map_meta_cap' ) );
+	}
+
+	/**
+	 * Test standard meta-caps.
+	 */
+	public function test_standard_meta_caps() {
+		foreach ( $this->generate_standard_meta_cap_tests() as $key => $test ) {
+			list ( $expecting, $user, $cap, $comment ) = $test;
+			$this->assertSame( $expecting, $user->has_cap( $cap, $comment ? $comment->comment_ID : null ), $key );
+		}
+	}
+
+	/**
+	 * Test meta-caps that check a post ID.
+	 */
+	public function test_post_check_meta_caps() {
+		foreach ( $this->generate_post_check_meta_cap_tests() as $key => $test ) {
+			list ( $expecting, $user, $cap, $post, $type ) = $test;
+			$this->assertSame( $expecting, $user->has_cap( $cap, $post ? $post->ID : null, $type ), $key );
+		}
+	}
+
+	/**
+	 * Test other pseudo-caps.
+	 */
+	public function test_other_pseudo_caps() {
+		foreach ( $this->generate_other_pseudo_cap_tests() as $key => $test ) {
+			list ( $expecting, $user, $cap, $type ) = $test;
+			$this->assertSame( $expecting, $user->has_cap( $cap, $type ), $key );
+		}
+	}
+
+	/*
+	 * --- Test generators -------------------------------------------------------
+	 */
+
+	/**
+	 * Generates tests for {@see test_standard_meta_caps()}.
+	 *
+	 * @return array Test arrays.
+	 */
+	protected function generate_standard_meta_cap_tests() {
```

Comment:
Do you know about [`@dataProvider`](https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers)? This makes PHPUnit aware of the data set and tests will continue if one of the data points fails.

Reply:
Thank you. I experimented with data providers a bit in the unit tests for this API. One thing I have against them is that they slow down unit tests that cover 1000's of permutations. Slower, because each dataset delivered by a data provider results in `setUp()` and `tearDown()` being called on each of those iterations; e.g.,  45k permutations x 2 setup/teardown calls.

I found that tests ran much faster if they were generated in logical sets, with groups of permutations being run together. Resulting in just one `setUp()` and `tearDown()`, which is all that was needed in most of the tests for this API.

Another thing I noticed is that all data providers run before any testing begins. So if a data-provider depends on any database-driven objects, those are lost when tests run, given all the heavy lifting done in `WP_UnitTestCase`.

Reply:
Ah, then it is not feasible.